HARK/ConsumptionSaving/ConsLabeledModel.py

```diff
@@ -0,0 +1,663 @@
+from dataclasses import dataclass
+from types import SimpleNamespace
+
+import numpy as np
+import xarray as xr
+from HARK.ConsumptionSaving.ConsIndShockModel import (
+    ConsIndShockSetup,
+    IndShockConsumerType,
+    init_perfect_foresight,
+)
+from HARK.ConsumptionSaving.ConsRiskyAssetModel import (
+    RiskyAssetConsumerType,
+    init_risky_asset,
+)
+from HARK.core import MetricObject, make_one_period_oo_solver
+from HARK.distribution import DiscreteDistributionLabeled
+from HARK.utilities import CRRAutility, CRRAutility_inv, CRRAutilityP, CRRAutilityP_inv
+
+
+class ValueFuncCRRALabeled(MetricObject):
+    def __init__(self, dataset: xr.Dataset, CRRA: float):
+
+        self.dataset = dataset
+        self.CRRA = CRRA
+
+    def __call__(self, state):
+        """
+        Interpolate inverse falue function then invert to get value function at given state.
+        """
+
+        if isinstance(state, xr.Dataset):
+            state_dict = state.coords
+        else:
+            state_dict = {}
+            for coords in self.dataset.coords.keys():
+                state_dict[coords] = state[coords]
+
+        result = CRRAutility(
+            self.dataset["v_inv"].interp(
+                state_dict,
+                assume_sorted=True,
+                kwargs={"fill_value": "extrapolate"},
+            ),
+            self.CRRA,
+        )
+
+        result.name = "v"
+        result.attrs = self.dataset["v"].attrs
+
+        return result
+
+    def derivative(self, state):
+        """
+        Interpolate inverse marginal value function then invert to get marginal value function at given state.
+        """
+
+        if isinstance(state, xr.Dataset):
+            state_dict = state.coords
+        else:
+            state_dict = {}
+            for coords in self.dataset.coords.keys():
+                state_dict[coords] = state[coords]
+
+        result = CRRAutilityP(
+            self.dataset["v_der_inv"].interp(
+                state_dict,
+                assume_sorted=True,
+                kwargs={"fill_value": "extrapolate"},
+            ),
+            self.CRRA,
+        )
+
+        result.name = "v_der"
+        result.attrs = self.dataset["v"].attrs
+
+        return result
+
+    def evaluate(self, state):
+        """
+        Interpolate all data variables in the dataset.
+        """
+
+        if isinstance(state, xr.Dataset):
+            state_dict = state.coords
+        else:
+            state_dict = {}
+            for coords in self.dataset.coords.keys():
+                state_dict[coords] = state[coords]
+
+        result = self.dataset.interp(
+            state_dict,
+            kwargs={"fill_value": "extrapolate"},
+        )
+        result.attrs = self.dataset["v"].attrs
+
+        return result
+
+
+class ConsumerSolutionLabeled(MetricObject):
+    def __init__(
+        self,
+        value: ValueFuncCRRALabeled,
+        policy: xr.Dataset,
+        continuation: ValueFuncCRRALabeled,
```

Comment:
To the extent that `value`, `policy`, and `continuation` all are wrappers around an underlying Dataset with shared coordinates, maybe this can be designed more flexibly.